### PR TITLE
THREESCALE-9976 Fix GRPC

### PR DESCRIPTION
**Pull request by eguzki**
### What

Fixes https://issues.redhat.com/browse/THREESCALE-9976

GRPC leverages the POST HTTP2 method ([GRPC Spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md)) . APIcast [reads the body](https://github.com/3scale/APIcast/pull/1419/files#diff-b4a0139162c96d858c01c0872560f495c15fee4e54df8ea52acda64e429c96cfR72) looking for request params for POST based requests using [ngx.req.read_body](https://openresty-reference.readthedocs.io/en/latest/Lua_Nginx_API/#ngxreqread_body). However, [ngx.req.read_body](https://openresty-reference.readthedocs.io/en/latest/Lua_Nginx_API/#ngxreqread_body) does not support HTTP2 and it blocks the request. The clients receive connection timeout errors.

Added development environment for GRPC scenario

### Verification steps

* Build docker image from this git branch

```shell
make runtime-image IMAGE_NAME=apicast-test
```
* Run grpc dev environment

```shell
cd dev-environments/grpc
make gateway-certs
make upstream-certs
make gateway IMAGE_NAME=apicast-test
```

* Download GRPC client

```shell
make grpcurl
```

* Run request to echo endpoint

```shell
bin/grpcurl -vv -insecure -H "app_id: abc123" -H "app_key: abc123" -authority gateway.example.com 127.0.0.1:8443 main.HelloWorld/Greeting
```

The command should return the echo response

```
Resolved method descriptor:
rpc Greeting ( .main.GreetingMessage ) returns ( .main.GreetingReply );

Request metadata to send:
app_id: abc123
app_key: abc123

Response headers received:
content-type: application/grpc
date: Fri, 03 Nov 2023 11:41:16 GMT
server: openresty

Estimated response size: 32 bytes

Response contents:
{
  "hostname": "d7728ab86df0",
  "clientAddress": "172.19.0.3:53564"
}

Response trailers received:
(empty)
Sent 0 requests and received 1 response
```

### CODE REVIEW

File: gateway/src/apicast/configuration/service.lua [hunk 1]
```diff
@@ -30,13 +33,42 @@ local http_methods_with_body = {
   PATCH = true
 }
 
+local function is_http2_request()
+  return ngx.req.http_version == '2.0'
+end
+
+local function content_type_is_urlencoded(headers)
+    local ct = headers["Content-Type"]
```
Comment (tkan145) [inline gateway/src/apicast/configuration/service.lua:41]: As I remember when running on Openshift, headers are all lower case. so perhaps use `ngx.var.http_content_type` instead?

File: dev-environments/grpc/Makefile [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+# go-install-tool will 'go install' any package $2 and install it to $1.
+define go-install-tool
+@[ -f $(1) ] || { \
+set -e ;\
+TMP_DIR=$$(mktemp -d) ;\
+cd $$TMP_DIR ;\
+go mod init tmp ;\
+echo "Downloading $(2)" ;\
+GOBIN=$(WORKDIR)/bin go install $(2) ;\
+rm -rf $$TMP_DIR ;\
```
Comment (tkan145) [inline dev-environments/grpc/Makefile:64]: What if user don't have Go installed on the machine? Perhaps we should put a prerequisite in the Readme file?

File: gateway/src/apicast/configuration/service.lua [hunk 3]
```diff
@@ -30,13 +33,42 @@ local http_methods_with_body = {
...
+    end
+
+    return str_find(lower(ct), "application/x-www-form-urlencoded", 1, true) ~= nil
+end
+
+
 local function read_body_args(...)
   local method = ngx.req.get_method()
 
+  ngx.log(ngx.WARN, "is_http2_request: ", is_http2_request())
```
Comment (tkan145) [inline gateway/src/apicast/configuration/service.lua:59]: I would move this log inside the if statement, otherwise this will print every time this function is called.

File: gateway/src/apicast/configuration/service.lua [hunk 4]
```diff
@@ -174,9 +206,13 @@ end
 local function get_request_params(method)
   local params = ngx.req.get_uri_args()
 
-  if method == "GET" then
+  ngx.log(ngx.WARN, "is_http2_request: ", is_http2_request())
```
Comment (tkan145) [inline gateway/src/apicast/configuration/service.lua:209]: Same as above

**Comment by tkan145** [comment]
Not sure why GH print the review twice :confused:

File: gateway/src/apicast/configuration/service.lua (hunk 3 above)
Comment (eguzki) [reply gateway/src/apicast/configuration/service.lua:59]: good catch! I forgot to remove it

File: gateway/src/apicast/configuration/service.lua (hunk 1 above)
Comment (eguzki) [reply gateway/src/apicast/configuration/service.lua:41]: From https://github.com/openresty/lua-nginx-module/tree/v0.10.19#ngxreqhttp_version

```
Also, by default, an __index metamethod is added to the resulting Lua table and will normalize the keys to a pure lowercase form with all underscores converted to dashes in case of a lookup miss. For example, if a request header My-Foo-Header is present, then the following invocations will all pick up the value of this header correctly:
 ngx.say(headers.my_foo_header)
 ngx.say(headers["My-Foo-Header"])
 ngx.say(headers["my-foo-header"])
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -22,11 +24,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - In boot mode on `init_worker` check configuration expiration [PR #1399](https://github.com/3scale/APIcast/pull/1399) [THREESCALE-9003](https://issues.redhat.com/browse/THREESCALE-9003)
 - Removes the warning message at the bootstrap [PR #1398](https://github.com/3scale/APIcast/pull/1398) [THREESCALE-7942](https://issues.redhat.com/browse/THREESCALE-7942)
 - Set NGiNX variable variables_hash_max_size to 2048 to avoid startup warning [PR #1395](https://github.com/3scale/APIcast/pull/1395) [THREESCALE-7941](https://issues.redhat.com/browse/THREESCALE-7941)
-- Dev environment on aarch64 host [PR #1381](https://github.com/3scale/APIcast/pull/1381)
```
Comment (tkan145) [inline CHANGELOG.md:25]: It's just a small thing but I think this is an unrelated change?

File: CHANGELOG.md [hunk 6]
```diff
@@ -22,11 +24,11 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 - In boot mode on `init_worker` check configuration expiration [PR #1399](https://github.com/3scale/APIcast/pull/1399) [THREESCALE-9003](https://issues.redhat.com/browse/THREESCALE-9003)
 - Removes the warning message at the bootstrap [PR #1398](https://github.com/3scale/APIcast/pull/1398) [THREESCALE-7942](https://issues.redhat.com/browse/THREESCALE-7942)
 - Set NGiNX variable variables_hash_max_size to 2048 to avoid startup warning [PR #1395](https://github.com/3scale/APIcast/pull/1395) [THREESCALE-7941](https://issues.redhat.com/browse/THREESCALE-7941)
-- Dev environment on aarch64 host [PR #1381](https://github.com/3scale/APIcast/pull/1381) 
+- Dev environment on aarch64 host [PR #1381](https://github.com/3scale/APIcast/pull/1381)
 
 ### Added
 
-- Doc: Policy Development Tutorial [PR #1384](https://github.com/3scale/APIcast/pull/1384)
```
Comment (tkan145) [inline CHANGELOG.md:29]: Same as above

**Comment by tkan145** [comment]
> However, [ngx.req.read_body](https://openresty-reference.readthedocs.io/en/latest/Lua_Nginx_API/#ngxreqread_body) does not support HTTP2 and it blocks the request.

Do you mean ngx.req.get_post_args? As the line `client request body saved to a temporary file` can be seen from the log, so I think ngx.req.read_body works with http2

File: CHANGELOG.md (hunk 5 above)
Comment (eguzki) [reply CHANGELOG.md:25]: It's just the IDE tool removing trailing spaces

File: CHANGELOG.md (hunk 6 above)
Comment (eguzki) [reply CHANGELOG.md:29]: It's just the IDE tool removing trailing spaces

**Review by tkan145 — APPROVED** [review]

**Comment by eguzki** [comment]
> > However, [ngx.req.read_body](https://openresty-reference.readthedocs.io/en/latest/Lua_Nginx_API/#ngxreqread_body) does not support HTTP2 and it blocks the request.
> 
> ~Do you mean ngx.req.get_post_args? As the line `client request body saved to a temporary file` can be seen from the log, so I think ngx.req.read_body works with http2~
> 
> Ignore my previous comment. just found this one [openresty/lua-nginx-module#2172](https://github.com/openresty/lua-nginx-module/issues/2172)

From https://github.com/openresty/lua-nginx-module#ngxreqread_body

```
Due to the stream processing feature of HTTP/2 or HTTP/3, this api could potentially block the entire request. Therefore, this api is effective only when HTTP/2 or HTTP/3 requests send content-length header. For requests with versions lower than HTTP/2, this api can still be used without any problems.
```
